### PR TITLE
[inductor] disable assert for indirect indexing

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120765

In https://github.com/pytorch/pytorch/issues/120452 we found negative indices handling together with the assertion for indirect indexing can cause perf loss. @htyu also mentioned that the device assert itself can slow things down.

I don't know how much degree of perf loss is caused by the device assert. Sending this PR so I can do a perf test.

### CODE REVIEW

**Comment by shunting314** [comment]
Some of our tests fails because they assumes we codegen a device-assert. These are easy to fix. But
[perf test](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2021%20Feb%202024%2018%3A57%3A29%20GMT&stopTime=Wed%2C%2028%20Feb%202024%2018%3A57%3A29%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/105/head&lCommit=a5864bfae049d665aeb433b7d7dd85dfe0e71f09&rBranch=main&rCommit=cbbc309caef049522cdebce981d229493539f681) result is neutral. 

I'll drop this PR unless we have other strong reason to proceed. @htyu
